### PR TITLE
Add unit tests for downloader models

### PR DESCRIPTION
**Pull request by harshsomankar123-tech**
fixes : #4738

### CODE REVIEW

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech Thank you for adding the test cases. I will have a look at this once I finish the https://github.com/kiwix/kiwix-android/pull/4747. I am stabilizing the test cases there. In the meantime, you can resolve the changes suggested by Copilot.

**Comment by kelson42** [comment]
I have rebased here as the CI was not passing... but we still need a proper code review.

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/ChunkUtilsTest.kt [hunk 1]
```diff
@@ -177,4 +177,31 @@ class ChunkUtilsTest {
...
+    assertEquals(5, listReturned[0].notificationID)
+  }
+
+  @Test
+  fun `getChunks chunk sizes sum up to at least contentLength`() {
+    val contentLength = ChunkUtils.CHUNK_SIZE * 2 + 500L
+    val listReturned = ChunkUtils.getChunks(url, contentLength, 1)
+    val totalSize = listReturned.sumOf { it.size }
+    assertThat(totalSize).isGreaterThanOrEqualTo(contentLength)
+  }
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/ChunkUtilsTest.kt:287]: We should test these edge cases as well.

```kotlin

  @Test
  fun `getChunks generates filenames correctly when file has no extension`() {
    every { StorageUtils.getFileNameFromUrl("noExtURL") } returns "TestFile"

    val size = ChunkUtils.CHUNK_SIZE * 2
    val chunks = ChunkUtils.getChunks("noExtURL", size, 10)

    assertEquals("TestFile.zimaa.part.part", chunks[0].fileName)
    assertEquals("TestFile.zimab.part.part", chunks[1].fileName)
  }

  @Test
  fun `chunk file extension rolls over after 26 chunks`() {
    val size = ChunkUtils.CHUNK_SIZE * 30
    val chunks = ChunkUtils.getChunks(url, size, 1)

    assertEquals("TestFileName.zimaa.part.part", chunks[0].fileName)
    assertEquals("TestFileName.zimaz.part.part", chunks[25].fileName)
    assertEquals("TestFileName.zimba.part.part", chunks[26].fileName)
  }

  @Test
  fun `chunks should have correct start and end byte`() {
    val size = ChunkUtils.CHUNK_SIZE * 2
    val chunks = ChunkUtils.getChunks(url, size, 1)

    assertEquals(0, chunks[0].startByte)
    assertEquals(ChunkUtils.CHUNK_SIZE, chunks[0].endByte)

    assertEquals(ChunkUtils.CHUNK_SIZE + 1, chunks[1].startByte)
  }

  @Test
  fun `last chunk endByte should match contentLength`() {
    val contentLength = ChunkUtils.CHUNK_SIZE + 500
    val chunks = ChunkUtils.getChunks(url, contentLength, 1)

    val lastChunk = chunks.last()

    assertThat(lastChunk.endByte).isEqualTo(contentLength)
  }

  @Test
  fun `getChunks works when url is null`() {
    every { StorageUtils.getFileNameFromUrl(null) } returns "Fallback"

    val chunks = ChunkUtils.getChunks(null, ChunkUtils.CHUNK_SIZE, 1)

    assertEquals("Fallback.part.part", chunks[0].fileName)
  }

  @Test
  fun `content length zero should produce no chunks`() {
    val chunks = ChunkUtils.getChunks(url, 0, 1)

    assertThat(chunks).isEmpty()
  }

  @Test
  fun `range header should include end when chunk not last`() {
    val size = ChunkUtils.CHUNK_SIZE * 3
    val chunks = ChunkUtils.getChunks(url, size, 1)

    assertEquals("0-${ChunkUtils.CHUNK_SIZE}", chunks[0].rangeHeader)
  }
```

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/ChunkUtilsTest.kt [hunk 2]
```diff
@@ -177,4 +177,31 @@ class ChunkUtilsTest {
...
+    assertEquals("0-", listReturned[0].rangeHeader)
+    assertEquals(5, listReturned[0].notificationID)
+  }
+
+  @Test
+  fun `getChunks chunk sizes sum up to at least contentLength`() {
+    val contentLength = ChunkUtils.CHUNK_SIZE * 2 + 500L
+    val listReturned = ChunkUtils.getChunks(url, contentLength, 1)
+    val totalSize = listReturned.sumOf { it.size }
+    assertThat(totalSize).isGreaterThanOrEqualTo(contentLength)
```
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/test/java/org/kiwix/kiwixmobile/core/downloader/ChunkUtilsTest.kt]: Apply this suggestion as well.

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/ChunkTest.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+package org.kiwix.kiwixmobile.core.downloader
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class ChunkTest {
+  @Test
+  fun `size returns 1 plus endByte minus startByte`() {
+    val chunk = Chunk(
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/ChunkTest.kt]: We are calling this `Chunk` constructor in each test. We should create a function for that.

Something like:

```kotlin
class ChunkTest {
  private fun createChunk(
    start: Long = 0L,
    end: Long = 99L,
    range: String = "$start-$end",
    fileName: String? = "test.zim.part.part",
    url: String? = "http://example.com/test.zim",
    contentLength: Long = end + 1,
    notificationID: Int = 1
  ) = Chunk(
    rangeHeader = range,
    fileName = fileName,
    url = url,
    contentLength = contentLength,
    notificationID = notificationID,
    startByte = start,
    endByte = end
  )

  @Test
  fun `constructor stores all properties correctly`() {
    val chunk = createChunk(
      start = 100,
      end = 200,
      range = "100-200",
      fileName = "wiki.zimaa.part.part",
      url = "http://mirror.example.com/wiki.zim",
      contentLength = 5000,
      notificationID = 42
    )

    assertThat(chunk.rangeHeader).isEqualTo("100-200")
    assertThat(chunk.fileName).isEqualTo("wiki.zimaa.part.part")
    assertThat(chunk.url).isEqualTo("http://mirror.example.com/wiki.zim")
    assertThat(chunk.contentLength).isEqualTo(5000L)
    assertThat(chunk.notificationID).isEqualTo(42)
    assertThat(chunk.startByte).isEqualTo(100L)
    assertThat(chunk.endByte).isEqualTo(200L)
  }

  @Test
  fun `constructor accepts null fileName and url`() {
    val chunk = createChunk(fileName = null, url = null)
    assertThat(chunk.fileName).isNull()
    assertThat(chunk.url).isNull()
  }

  @Test
  fun `size returns 1 plus endByte minus startByte`() {
    val chunk = createChunk(0, 99)
    assertThat(chunk.size).isEqualTo(100)
  }

  @Test
  fun `size is correct when startByte is non-zero`() {
    val chunk = createChunk(50, 149)
    assertThat(chunk.size).isEqualTo(100)
  }

  @Test
  fun `size is 1 when startByte equals endByte`() {
    val chunk = createChunk(0, 0)
    assertThat(chunk.size).isEqualTo(1)
  }

  @Test
  fun `size works for large ranges`() {
    val start = 0L
    val end = ChunkUtils.CHUNK_SIZE
    val chunk = createChunk(start, end)
    assertThat(chunk.size).isEqualTo(ChunkUtils.CHUNK_SIZE + 1)
  }

  @Test
  fun `size works near long boundaries`() {
    val start = Long.MAX_VALUE - 10
    val end = Long.MAX_VALUE - 1

    val chunk = createChunk(
      start = start,
      end = end,
      range = "$start-$end",
      contentLength = Long.MAX_VALUE
    )
    assertThat(chunk.size).isEqualTo(10)
  }

  @Test
  fun `size calculation still follows formula when startByte greater than endByte`() {
    val chunk = createChunk(100, 50)
    assertThat(chunk.size).isEqualTo(-49)
  }

  @Test
  fun `size returns consistent value across multiple accesses`() {
    val chunk = createChunk(0, 9)
    assertThat(chunk.size).isEqualTo(10)
    assertThat(chunk.size).isEqualTo(10)
    assertThat(chunk.size).isEqualTo(10)
  }

  @Test
  fun `isDownloaded defaults to false`() {
    val chunk = createChunk()
    assertThat(chunk.isDownloaded).isFalse()
  }

  @Test
  fun `isDownloaded can be set to true`() {
    val chunk = createChunk()
    chunk.isDownloaded = true
    assertThat(chunk.isDownloaded).isTrue()
  }

  @Test
  fun `isDownloaded can toggle between states`() {
    val chunk = createChunk()
    chunk.isDownloaded = true
    assertThat(chunk.isDownloaded).isTrue()

    chunk.isDownloaded = false
    assertThat(chunk.isDownloaded).isFalse()
  }

  @Test
  fun `size is correct for large byte ranges`() {
    val startByte = 0L
    val endByte = ChunkUtils.CHUNK_SIZE
    val chunk = createChunk(start = startByte, end = endByte)
    assertThat(chunk.size).isEqualTo(ChunkUtils.CHUNK_SIZE + 1)
  }
}
```

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/SecondsTest.kt [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.jupiter.api.Test
+import org.kiwix.kiwixmobile.core.CoreApp
+import org.kiwix.kiwixmobile.core.R
+
+class SecondsTest {
+  private lateinit var mockContext: Context
+
+  @BeforeEach
+  fun setUp() {
+    mockContext = io.mockk.mockk(relaxed = true)
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/SecondsTest.kt]: We should not import with package name :(

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/SecondsTest.kt [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    val seconds = Seconds(86400L)
+    assertThat(seconds.toHumanReadableTime()).isEqualTo("1 day left")
+  }
+
+  @Test
+  fun `toHumanReadableTime formats multiple days correctly`() {
+    val seconds = Seconds(172800L)
+    assertThat(seconds.toHumanReadableTime()).isEqualTo("2 day left")
+  }
+
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/SecondsTest.kt:69]: Here we are missing some scenarios.

```kotlin

  @Test
  fun `toHumanReadableTime rounds near one day to one day`() {
    val seconds = Seconds(86399L)
    assertThat(seconds.toHumanReadableTime()).isEqualTo("1 day left")
  }


  @Test
  fun `toHumanReadableTime rounds near one hour to one hour`() {
    val seconds = Seconds(3599L)
    assertThat(seconds.toHumanReadableTime()).isEqualTo("1 hour left")
  }

 @Test
  fun `toHumanReadableTime rounds near one minute to one minute`() {
    val seconds = Seconds(59L)
    assertThat(seconds.toHumanReadableTime()).isEqualTo("1 minute left")
  }

 @Test
  fun `handles very large durations`() {
    val seconds = Seconds(86400 * 10L)
    assertThat(seconds.toHumanReadableTime()).isEqualTo("10 day left")
  }

  @Test
  fun `context strings are retrieved for formatting`() {
    val seconds = Seconds(3600)
    seconds.toHumanReadableTime()

    verify { mockContext.getString(R.string.time_hour) }
    verify { mockContext.getString(R.string.time_left) }
  }
```

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  fun `toString returns Failed for Failed state`() {
+    val failed = DownloadState.Failed(Error.UNKNOWN, null)
+    assertThat(failed.toString()).isEqualTo("Failed(reason=UNKNOWN, zimUrl=null)")
+  }
+
+  @Test
+  fun `toReadableState returns correct string for Pending`() {
+    val context = mockk<Context>()
+    every { context.getString(R.string.pending_state) } returns "Pending"
+    assertThat(DownloadState.Pending.toReadableState(context)).isEqualTo("Pending")
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt:142]: After asserting, we should verify whether the `pending_state` string is called.

```kotlin
verify { context.getString(R.string.pending_state) }
```

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Test
+  fun `toReadableState returns correct string for Running`() {
+    val context = mockk<Context>()
+    every { context.getString(R.string.running_state) } returns "Downloading"
+    assertThat(DownloadState.Running.toReadableState(context)).isEqualTo("Downloading")
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt:150]: Same here.

 verify { context.getString(R.string.running_state) }

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Test
+  fun `toReadableState returns correct string for Paused`() {
+    val context = mockk<Context>()
+    every { context.getString(R.string.paused_state) } returns "Paused"
+    assertThat(DownloadState.Paused.toReadableState(context)).isEqualTo("Paused")
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt:158]: Here too

verify { context.getString(R.string.paused_state) }

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Test
+  fun `toReadableState returns correct string for Successful`() {
+    val context = mockk<Context>()
+    every { context.getString(R.string.complete) } returns "Complete"
+    assertThat(DownloadState.Successful.toReadableState(context)).isEqualTo("Complete")
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt:166]: verify { context.getString(R.string.complete) }

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Test
+  fun `toReadableState returns correct string for Failed`() {
+    val context = mockk<Context>()
+    val failed = DownloadState.Failed(Error.CONNECTION_TIMED_OUT, null)
+    every { context.getString(R.string.failed_state, "CONNECTION_TIMED_OUT") } returns
+      "Failed: CONNECTION_TIMED_OUT"
+    assertThat(failed.toReadableState(context)).isEqualTo("Failed: CONNECTION_TIMED_OUT")
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt:176]: Here too

verify {
      context.getString(R.string.failed_state, "CONNECTION_TIMED_OUT")
    }

File: core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+  }
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/downloader/model/DownloadStateTest.kt:188]: We should add this test:

```kotlin
 @Test
  fun `toReadableState works when Failed has zimUrl`() {
    val context = mockk<Context>()
    val failed = DownloadState.Failed(Error.UNKNOWN, "http://example.com")

    every { context.getString(R.string.failed_state, "UNKNOWN") } returns "Failed"

    assertThat(failed.toReadableState(context)).isEqualTo("Failed")
  }
```

**Comment by harshsomankar123-tech** [comment]
@MohitMaliFtechiz Thanks for reviewing. I’ve implemented all the suggestions.  
PTAL.

**Review by MohitMaliFtechiz — APPROVED** [review]
@harshsomankar123-tech Thank you 🎉 🎉 . @kelson42 LGTM.
